src/Libraries/Microsoft.Extensions.AI/ChatCompletion/OpenTelemetryImageGenerator.cs

```diff
@@ -0,0 +1,313 @@
+        _activitySource = new(name);
+        _meter = new(name);
+
+        _tokenUsageHistogram = _meter.CreateHistogram<int>(
+            OpenTelemetryConsts.GenAI.Client.TokenUsage.Name,
+            OpenTelemetryConsts.TokensUnit,
+            OpenTelemetryConsts.GenAI.Client.TokenUsage.Description
+#if NET9_0_OR_GREATER
+            , advice: new() { HistogramBucketBoundaries = OpenTelemetryConsts.GenAI.Client.TokenUsage.ExplicitBucketBoundaries }
+#endif
+            );
+
+        _operationDurationHistogram = _meter.CreateHistogram<double>(
+            OpenTelemetryConsts.GenAI.Client.OperationDuration.Name,
+            OpenTelemetryConsts.SecondsUnit,
+            OpenTelemetryConsts.GenAI.Client.OperationDuration.Description
+#if NET9_0_OR_GREATER
+            , advice: new() { HistogramBucketBoundaries = OpenTelemetryConsts.GenAI.Client.OperationDuration.ExplicitBucketBoundaries }
+#endif
+            );
+    }
+
+    /// <inheritdoc/>
+    protected override void Dispose(bool disposing)
+    {
+        if (disposing)
+        {
+            _activitySource.Dispose();
+            _meter.Dispose();
+        }
+
+        base.Dispose(disposing);
+    }
+
+    /// <summary>
+    /// Gets or sets a value indicating whether potentially sensitive information should be included in telemetry.
+    /// </summary>
+    /// <value>
+    /// <see langword="true"/> if potentially sensitive information should be included in telemetry;
+    /// <see langword="false"/> if telemetry shouldn't include raw inputs and outputs.
+    /// The default value is <see langword="false"/>, unless the <c>OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT</c>
+    /// environment variable is set to "true" (case-insensitive).
+    /// </value>
+    /// <remarks>
+    /// By default, telemetry includes metadata, such as token counts, but not raw inputs
+    /// and outputs, such as message content, function call arguments, and function call results.
+    /// The default value can be overridden by setting the <c>OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT</c>
+    /// environment variable to "true". Explicitly setting this property will override the environment variable.
+    /// </remarks>
+    public bool EnableSensitiveData { get; set; } = TelemetryHelpers.EnableSensitiveDataDefault;
+
+    /// <inheritdoc/>
+    public override object? GetService(Type serviceType, object? serviceKey = null) =>
+        serviceType == typeof(ActivitySource) ? _activitySource :
+        base.GetService(serviceType, serviceKey);
+
+    /// <inheritdoc/>
+    public async override Task<ImageGenerationResponse> GenerateAsync(
+        ImageGenerationRequest request, ImageGenerationOptions? options = null, CancellationToken cancellationToken = default)
+    {
+        _ = Throw.IfNull(request);
+
+        using Activity? activity = CreateAndConfigureActivity(request, options);
+        Stopwatch? stopwatch = _operationDurationHistogram.Enabled ? Stopwatch.StartNew() : null;
```

Comment:
Is this the standard way to measure activity duration in otel? I wonder if there should be an API that pools stopwatch objects.

Reply:
While it's a bit antithetical for me in particular, I've not been terribly concerned about such allocation when it makes the code more convoluted; the amount of other overhead in one of these operations dwarfs a 24 byte temporary object allocation.

Once this repo updates to C# 14 and we can polyfill Stopwatch.GetElapsedTime for netstandard/netfx, changing this to use GetTimestamp and GetElapsedTime will be a fine thing to do.